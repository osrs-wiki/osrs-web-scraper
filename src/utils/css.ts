import { HTMLElement } from "node-html-parser";

/**
 * Extracts background image URL from CSS style attribute
 * Handles both single and double quotes, and HTML entity encoding
 */
export function extractBackgroundImageUrl(style: string): string | null {
  if (!style) return null;
  
  // First decode HTML entities
  const decodedStyle = style
    .replace(/&quot;/g, '"')
    .replace(/&apos;/g, "'")
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&amp;/g, '&');
  
  // Match background-image: url("...") or url('...') patterns
  const urlMatch = decodedStyle.match(/background-image\s*:\s*url\s*\(\s*['"']?([^'"')]+)['"']?\s*\)/i);
  
  if (urlMatch && urlMatch[1]) {
    return urlMatch[1];
  }
  
  return null;
}

/**
 * Extracts all background image URLs from style attributes in HTML elements
 */
export function extractBackgroundImages(elements: HTMLElement[]): string[] {
  const imageUrls: string[] = [];
  
  elements.forEach(element => {
    if (element.attributes?.style) {
      const url = extractBackgroundImageUrl(element.attributes.style);
      if (url) {
        imageUrls.push(url);
      }
    }
  });
  
  return imageUrls;
}