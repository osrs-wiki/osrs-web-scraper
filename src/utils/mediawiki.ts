import { MediaWikiBreak, MediaWikiContent } from "@osrs-wiki/mediawiki-builder";
/**
 * Trims the leading and trailing MediaWikiBreak elements from the contents.
 *
 * @param contents The contents to trim.
 * @returns The trimmed contents.
 */
export const trim = (contents: MediaWikiContent[]): MediaWikiContent[] => {
  if (!contents || contents.length === 0) {
    return contents;
  }
  while (
    contents[0] instanceof MediaWikiBreak ||
    contents[contents.length - 1] instanceof MediaWikiBreak
  ) {
    if (contents.length > 0 && contents[0] instanceof MediaWikiBreak) {
      contents.shift();
    }
    if (
      contents.length > 0 &&
      contents[contents.length - 1] instanceof MediaWikiBreak
    ) {
      contents.pop();
    }
  }
  return contents;
};

/**
 * Checks if the contents are empty.
 * @param contents The contents to check.
 * @returns True if the contents are empty, otherwise false.
 */
export const isEmpty = (
  contents: string | MediaWikiContent | MediaWikiContent[]
): boolean => {
  if (typeof contents === "string") {
    return contents.trim().length === 0;
  } else if (Array.isArray(contents)) {
    return contents.length === 0 || contents.every(isEmpty);
  } else if (contents instanceof MediaWikiContent) {
    return isEmpty(contents.children);
  }
  return false;
};

/**
 * Checks if the first child of the contents starts with the given string.
 *
 * @param contents The contents to check.
 * @param str The string to check against.
 * @returns True if the first child starts with the given string, otherwise false.
 */
export const startsWith = (
  contents: string | MediaWikiContent | MediaWikiContent[],
  str: string
): boolean => {
  if (isEmpty(contents)) {
    return false;
  }
  if (typeof contents === "string") {
    return contents.startsWith(str);
  }
  const childContents = Array.isArray(contents) ? contents[0] : contents;
  if (typeof childContents.children === "string") {
    return childContents.children.startsWith(str);
  }
  return startsWith(childContents.children, str);
};

/**
 * Recursively retrieves the first string content from the given MediaWikiContent.
 *
 * @param contents The MediaWikiContent to search through.
 * @returns The first string content found, or undefined if none is found.
 */
export const getFirstStringContent = (
  contents: MediaWikiContent
): MediaWikiContent | undefined => {
  if (typeof contents.children === "string") {
    return contents;
  } else if (contents.children instanceof MediaWikiContent) {
    return getFirstStringContent(contents.children);
  } else if (Array.isArray(contents.children) && contents.children.length > 0) {
    return getFirstStringContent(contents.children[0]);
  }
  return undefined;
};
