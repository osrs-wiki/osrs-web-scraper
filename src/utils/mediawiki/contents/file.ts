import MediaWikiContent from "../content";

type MediaWikiFileOptions = {
  format?: "frameless" | "frame" | "thumb";
  resizing?: {
    width?: number;
    height?: number;
  };
  horizontalAlignment?: "left" | "right" | "center" | "none";
  verticalAlignment?:
    | "baseline"
    | "sub"
    | "super"
    | "top"
    | "text-top"
    | "bottom"
    | "text-bottom";
  link?: string;
};

class MediaWikiFile extends MediaWikiContent {
  fileName: string;
  options?: MediaWikiFileOptions;

  constructor(fileName: string, options?: MediaWikiFileOptions) {
    super();
    this.fileName = fileName;
    this.options = options;
  }

  build() {
    let options = "";
    const { format, resizing, horizontalAlignment, verticalAlignment, link } =
      this.options;
    if (format) {
      options += `|${format}`;
    }
    if (resizing?.width && resizing?.height) {
      options += `|${resizing.width}x${resizing.height}px`;
    } else if (resizing?.width) {
      options += `|${resizing.width}px`;
    } else if (resizing?.height) {
      options += `|${resizing.height}px`;
    }
    if (horizontalAlignment) {
      options += `|${horizontalAlignment}`;
    }
    if (verticalAlignment) {
      options += `|${verticalAlignment}`;
    }
    if (link) {
      options += `|link=${link}`;
    }
    return `[[File:${this.fileName}${options}]]`;
  }
}

export default MediaWikiFile;
