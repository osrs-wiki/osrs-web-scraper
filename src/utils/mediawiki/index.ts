export { default as MediaWikiBuilder } from "./builder";
export { default as MediaWikiContent } from "./content";

export * from "./contents";
