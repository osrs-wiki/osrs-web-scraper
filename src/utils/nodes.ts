import { Node } from "node-html-parser";

/**
 * Util function to prevent having to include ts-ignore every time rawTagName is referenced.
 * @param node
 * @returns
 */
export const getNodeTagName = (node: Node): string => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  return node.rawTagName;
};
