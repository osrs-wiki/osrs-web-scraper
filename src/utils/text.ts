/**
 * Format text content for parsing.
 * @param text
 * @returns
 */
export const formatText = (text: string) =>
  text
    .replaceAll("&amp;", "&")
    .replaceAll("\t", "")
    .replaceAll("\n", "")
    .replaceAll("\r", "");
