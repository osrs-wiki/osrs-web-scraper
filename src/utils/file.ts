import fs from "fs";
import client from "https";
import path from "path";

import { formatText } from "./text";

/**
 * Download a file from a url
 * @param url The url of the file to download
 * @param filepath The filepath of the downloaded file
 * @returns
 */
export const downloadFile = async (url: string, filepath: string) => {
  console.info(`Attempting file download: ${url}`);
  return new Promise((resolve, reject) => {
    client.get(url, (res) => {
      if (res.statusCode === 200) {
        res
          .pipe(fs.createWriteStream(filepath))
          .on("error", reject)
          .once("close", () => resolve(filepath));
        console.info(`Downloaded file ${filepath}`);
      } else {
        res.resume();
        reject(
          new Error(`Request Failed With a Status Code: ${res.statusCode}`)
        );
      }
    });
  });
};

/**
 * Get a file from a url
 * @param fileUrl The file url
 * @returns
 */
export const getFileName = (fileUrl: string) => {
  const parsed = new URL(fileUrl);
  return path.basename(parsed.pathname);
};

/**
 * Get the extension of a file from a url link.
 * @param fileUrl The file url
 * @returns
 */
export const getFileExtension = (fileUrl: string) => {
  return fileUrl.split(/[#?]/)[0].split(".").pop().trim();
};

const fileCharsReplaceMap: { [key: string]: string } = {
  ":": "-",
  "|": "-",
  "&nbsp;": "",
};

/**
 * Format a file name to ensure it is valid.
 * @param fileName The file name
 * @returns
 */
export const formatFileName = (fileName: string) => {
  const cleanedFileName = Object.keys(fileCharsReplaceMap).reduce(
    (name, find) => name.replaceAll(find, fileCharsReplaceMap[find]),
    fileName
  );
  return formatText(cleanedFileName);
};
