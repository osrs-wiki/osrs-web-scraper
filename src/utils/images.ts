import fs from "fs";
import client from "https";
import path from "path";

import { formatText } from "./text";

/**
 * Download an image from a url
 * @param url The url of the image to download
 * @param filepath The filepath of the downloaded image
 * @returns
 */
export const downloadImage = async (url: string, filepath: string) => {
  console.info(`Attempting image download: ${url}`);
  return new Promise((resolve, reject) => {
    client.get(url, (res) => {
      if (res.statusCode === 200) {
        res
          .pipe(fs.createWriteStream(filepath))
          .on("error", reject)
          .once("close", () => resolve(filepath));
        console.info(`Downloaded image ${filepath}`);
      } else {
        res.resume();
        reject(
          new Error(`Request Failed With a Status Code: ${res.statusCode}`)
        );
      }
    });
  });
};

/**
 * Get an image name from a url
 * @param imageUrl The image url
 * @returns
 */
export const getImageName = (imageUrl: string) => {
  const parsed = new URL(imageUrl);
  return path.basename(parsed.pathname);
};

/**
 * Get the extension of an image from a url link.
 * @param imageUrl The image url
 * @returns
 */
export const getImageExtension = (imageUrl: string) => {
  return imageUrl.split(/[#?]/)[0].split(".").pop().trim();
};

/**
 * Format a file name to ensure it is valid.
 * @param fileName The file name
 * @returns
 */
export const formatFileName = (fileName: string) => {
  return formatText(fileName.replaceAll(":", "-"));
};
