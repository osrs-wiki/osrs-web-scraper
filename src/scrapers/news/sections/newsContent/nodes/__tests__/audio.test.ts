import { MediaWikiBuilder } from "@osrs-wiki/mediawiki-builder";
import parse from "node-html-parser";

import * as fileUtils from "../../../../../../utils/file";
import audioParser from "../audio";

describe("audio node", () => {
  test("audio node should parse and render", () => {
    const downloadFileSpy = jest
      .spyOn(fileUtils, "downloadFile")
      .mockImplementationOnce(() => Promise.resolve("./out/news/test title/test title narration.mp3"));

    const root = parse(
      '<audio controls=""><source src="test.mp3" type="audio/wav"></audio>'
    );
    const builder = new MediaWikiBuilder();
    builder.addContents(
      [audioParser(root.firstChild, { title: "test title" })].flat()
    );

    expect(downloadFileSpy).toHaveBeenCalledWith(
      "test.mp3",
      "./out/news/test title/test title narration.mp3"
    );
    expect(builder.build()).toMatchSnapshot();
  });
});
