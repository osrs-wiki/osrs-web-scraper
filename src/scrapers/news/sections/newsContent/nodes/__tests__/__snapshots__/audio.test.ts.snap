// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`audio node audio node should parse and render 1`] = `
"[[File:test title narration.mp3|center]]
"
`;
